packages/superset-ui-core/src/callApi/parseResponse.js

```diff
@@ -1,26 +1,29 @@
-export default function parseResponse(apiPromise) {
-  return apiPromise.then(apiResponse =>
+const PARSERS = {
+  json: response =>
     // first try to parse as json, and fall back to text (e.g., in the case of HTML stacktrace)
     // cannot fall back to .text() without cloning the response because body is single-use
-    apiResponse
+    response
       .clone()
       .json()
-      .catch(() => /* jsonParseError */ apiResponse.text().then(textPayload => ({ textPayload })))
-      .then(maybeJson => ({
-        json: maybeJson.textPayload ? undefined : maybeJson,
-        response: apiResponse,
-        text: maybeJson.textPayload,
-      }))
-      .then(({ response, json, text }) => {
-        if (!response.ok) {
-          return Promise.reject({
-            error: response.error || (json && json.error) || text || 'An error occurred',
-            status: response.status,
-            statusText: response.statusText,
-          });
-        }
+      .then(json => ({ json, response }))
+      .catch(() => /* jsonParseError */ response.text().then(text => ({ response, text }))),
 
-        return typeof text === 'undefined' ? { json, response } : { response, text };
-      }),
-  );
+  text: response => response.text().then(text => ({ response, text })),
+};
+
+export default function parseResponse(apiPromise, parseMethod = 'json') {
+  const responseParser = PARSERS[parseMethod] || PARSERS.json;
+
+  return apiPromise.then(responseParser).then(({ json, response, text }) => {
+    // HTTP 404 or 500 are not rejected, ok is just set to false
+    if (!response.ok) {
+      return Promise.reject({
+        error: response.error || (json && json.error) || text || 'An error occurred',
```

Comment:
Perhaps should separate error handling for `apiResponse`, `json` parsing and `text` parsing.

Reply:
```js
return apiPromise
  .then((response) => {
    if (!response.ok) {
      return Promise.reject({
        error: response.error,
        status: response.status,
        statusText: response.statusText,
      });
    }
    return response;
  })
  .then(responseParser)
```

Reply:
I dig the latter^